### DISCUSSION TITLE
Sharing of subterms

### DISCUSSION THREAD

**Discussion by bruderj15** [usi-verification-and-security/opensmt#773] [Q&A] [answered]
Hello. 
How does `OpenSMT` handle sharing of terms, especially regarding `let`-binders?

Z3 just expands the `let` because they share identical terms anyways: https://github.com/Z3Prover/z3/issues/6732#issuecomment-1560704954

**Comment by blishko** [comment, marked as answer]
Internally, each term in OpenSMT is unique. When it is shared as a subterm in multiple terms, each term only has a "reference" (`PTRef`) to the unique term.
I would say this is similar to Z3 (although the implementation details are different, I would say).
`lets` are only handled by the SMT-LIB frontend, they do not exist internally.
